### PR TITLE
IAT plugin

### PR DESCRIPTION
**Pull request by kristiyip**

### CODE REVIEW

**Comment by jodeleeuw** [comment]
@kristiyip Can you add the test file to this pull request as well, to make debugging easier? Also we'll want that file in the repo eventually!

In the latest version of jsPsych `jsPsych.ALL_KEYS` should be used in lieu of `[]` for representing any key. That might explain the problem?

**Comment by jodeleeuw** [comment]
Oh now I see that the example is there, it's just not following the naming convention. You could rename to `jspsych-iat.html`.

**Comment by jodeleeuw** [comment]
@kristiyip can you also implement a test file like https://github.com/jodeleeuw/jsPsych/blob/master/tests/plugins/plugin-single-stim.test.js but for this IAT plugin?

I am working towards getting all plugins properly tested so I'd like to get a test written for this one before the merge.

**Comment by jodeleeuw** [comment]
@kristiyip how's this going? do you need any feedback from me to make more progress on this?

**Comment by kristiyip** [comment]
Feedback would be great! I had updated some packages for testing but the tests still failed when I pushed the update. On my personal laptop the tests passed though. 


Sent via the Samsung Galaxy S7, an AT&T 4G LTE smartphone
-------- Original message --------From: Josh de Leeuw <notifications@github.com> Date: 5/26/17  12:07 AM  (GMT-05:00) To: jodeleeuw/jsPsych <jsPsych@noreply.github.com> Cc: kristiyip <krdiep@vassar.edu>, Mention <mention@noreply.github.com> Subject: Re: [jodeleeuw/jsPsych] IAT plugin (#364) 
@kristiyip how's this going? do you need any feedback from me to make more progress on this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/jodeleeuw/jsPsych","title":"jodeleeuw/jsPsych","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/jodeleeuw/jsPsych"}},"updates":{"snippets":[{"icon":"PERSON","message":"@jodeleeuw in #364: @kristiyip how's this going? do you need any feedback from me to make more progress on this?"}],"action":{"name":"View Pull Request","url":"https://github.com/jodeleeuw/jsPsych/pull/364#issuecomment-304185955"}}}

**Comment by jodeleeuw** [comment]
@kristiyip can you try merging the master branch into your branch, deleting the node_modules folder, running `npm install` and then run `npm test -- -u`? I think that will update jest to the latest version and then update the snapshot files. That should get it in sync with the CI server.

**Comment by kristiyip** [comment]
@jodeleeuw Checks passed!

File: README.md [hunk 1]
```diff
@@ -41,3 +41,4 @@ Credits
 jsPsych was created by Josh de Leeuw ([@jodeleeuw](https://github.com/jodeleeuw)).
 
 There have been many [contributors](https://github.com/jodeleeuw/jsPsych/blob/master/contributors.md) to the library. Thank you!
+# IATPlugin
```
Comment (jodeleeuw) [inline README.md]: Can you remove this header?

File: docs/markdown_docs/plugins/jspsych-iat.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
+# jspsych-iat plugin
+
+This plugin displays an image or HTML-formatted content that is the stimulus of the experiement. Each stimulus is assigned a category and a key that the user presses to give a response. The stimulus can be displayed until a response is given, or for a pre-determined amount of time. If the stimulus is displayed until a response is given, then an image to indicate if the user pressed the wrong key can be displayed. The trial can be ended automatically if the subject has failed to respond within a fixed length of time.
```
Comment (jodeleeuw) [inline docs/markdown_docs/plugins/jspsych-iat.md]: This description should be updated to convey how the IAT works (or provide a link to a reference on the IAT, perhaps)

File: docs/markdown_docs/plugins/jspsych-iat.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+In addition to the [default data collected by all plugins](overview#datacollectedbyplugins), this plugin collects the following data for each trial.
+
+Name | Type | Value
+-----|------|------
+stimulus | string | Either the path to the image file or the string containing the HTML formatted content that the subject saw on this trial.
+key_press | numeric | Indicates which key the subject pressed. The value is the [numeric key code](http://www.cambiaresearch.com/articles/15/javascript-char-codes-key-codes) corresponding to the subject's response.
+rt | numeric | The response time in milliseconds for the subject to make a response. The time is measured from when the stimulus first appears on the screen until the subject's response.
+correct | boolean | Boolean of whether the user's key press was correct for the given image or incorrect. 
+
+## Examples
```
Comment (jodeleeuw) [inline docs/markdown_docs/plugins/jspsych-iat.md]: Can the example here be shorter? The longer example is great, but probably not appropriate in the docs. You could put a link to the longer example in here.

File: plugins/jspsych-iat.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
+/**
+ * jspsych-single-stim
```
Comment (jodeleeuw) [inline plugins/jspsych-iat.js]: change name here

File: plugins/jspsych-iat.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    trial.wrong_image_name = trial.wrong_image_name || 'wrong image';
+    trial.left_category_key = trial.left_category_key || 'E';
+    trial.right_category_key = trial.right_category_key || 'I';
+    trial.left_category_label = trial.left_category_label || ['left'];
+    trial.right_category_label = trial.right_category_label || ['right']; 
+    trial.stim_key_association = trial.stim_key_association || 'undefined';
+    trial.response_ends_trial = (typeof trial.response_ends_trial == 'undefined') ? true : trial.response_ends_trial;
+    trial.timing_response = trial.timing_response || -1;
+    trial.key_to_move_forward = trial.key_to_move_forward || [jsPsych.ALL_KEYS];
+    trial.is_html = (typeof trial.is_html == 'undefined') ? false : trial.is_html;
```
Comment (jodeleeuw) [inline plugins/jspsych-iat.js]: Are we allowing HTML content for the stimulus?

File: plugins/jspsych-iat.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+ * documentation: docs.jspsych.org
+ *
+ **/
+
+
+ jsPsych.plugins["iat"] = (function() {
+
+  var plugin = {};
+
+  jsPsych.pluginAPI.registerPreload('iat', 'stimulus', 'image');
```
Comment (jodeleeuw) [inline plugins/jspsych-iat.js]: If we are allowing HTML content for stimulus, this needs to change slightly. See the example in single-stim for the fourth parameter that needs to get added.

File: docs/markdown_docs/plugins/jspsych-iat.md (hunk 2 above)
Comment (kristiyip) [inline docs/markdown_docs/plugins/jspsych-iat.md]: If I provide a link, should it be https://implicit.harvard.edu/implicit/takeatest.html ?

Comment (jodeleeuw) [inline docs/markdown_docs/plugins/jspsych-iat.md]: I'd suggest https://implicit.harvard.edu/implicit/iatdetails.html

File: docs/markdown_docs/plugins/jspsych-iat.md (hunk 3 above)
Comment (kristiyip) [inline docs/markdown_docs/plugins/jspsych-iat.md]: Alright so maybe instead have one trial block?

File: plugins/jspsych-iat.js (hunk 5 above)
Comment (kristiyip) [inline plugins/jspsych-iat.js]: I thought we were for text stimulus?

Comment (jodeleeuw) [inline plugins/jspsych-iat.js]: Right, good point.

File: docs/markdown_docs/plugins/jspsych-iat.md (hunk 3 above)
Comment (jodeleeuw) [inline docs/markdown_docs/plugins/jspsych-iat.md]: Yes, or even just a single trial. That's how most of these examples work in the documentation.

**Pull request merged**